### PR TITLE
Add fanout storage which can query multiple storages in parallel

### PR DESCRIPTION
**Pull request by nikunjgit**

### CODE REVIEW

File: storage/fanout/storage.go [hunk 1]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
...
+		fetchResult, ok := res.Value.(*storage.FetchResult)
+		if !ok {
+			return nil, fmt.Errorf("invalid fetch result")
+		}
+
+		if fetchreq.store.Type() != storage.TypeLocalDC {
+			result.LocalOnly = false
+		}
+
+		result.SeriesList = append(result.SeriesList, fetchResult.SeriesList...)
```
Comment (arnikola) [inline storage/fanout/storage.go]: `result.SeriesList[i] = result.SeriesList`?

File: util/execution/orderedParallel.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	Err   error
+	Order int
+}
+
+// ExecuteParallel executes a slice of requests in parallel and returns ordered results
+func ExecuteParallel(requests []Request) []*Response {
+	unordered := make(chan *Response)
+	go processParallel(requests, unordered)
+	return orderResults(unordered, len(requests))
+
```
Comment (arnikola) [inline util/execution/orderedParallel.go]: unnecessary whitespace

File: util/execution/orderedParallel.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+func processParallel(requests []Request, responses chan<- *Response) {
+	defer close(responses)
+	var wg sync.WaitGroup
+	wg.Add(len(requests))
+	for _, req := range requests {
+		go func(r Request) {
+			responses <- r.Process()
```
Comment (arnikola) [inline util/execution/orderedParallel.go]: Not sure if this is standard, but most of the examples I've seen do it like this:

```golang
... {
 defer wg.Done()
 responses <- r.Process()
} ...
```

File: storage/fanout/storage.go [hunk 4]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
...
-	return stores[0].Write(ctx, query)
+	requests := make([]execution.Request, len(stores))
+	for idx, store := range stores {
+		requests[idx] = newWriteRequest(ctx, store, query, idx)
+	}
+
+	responses := execution.ExecuteParallel(requests)
+	for _, resp := range responses {
+		if resp.Err != nil {
+			return resp.Err
```
Comment (arnikola) [inline storage/fanout/storage.go]: I think this might lead to incorrect behaviour; if one write out of 10 stores fails, this will return an error for the full write, and retries will write the value twice to 9 stores

File: util/execution/orderedParallel.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+			wg.Done()
+		}(req)
+	}
+	wg.Wait()
+}
+
+func orderResults(unordered <- chan *Response, size int) []*Response {
+	orderedResponses := make([]*Response, size)
```
Comment (arnikola) [inline util/execution/orderedParallel.go]: A bit of a stretch, but there's nothing enforcing that `requests.order` are well-formed increasing natural numbers

File: storage/fanout/storage_test.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	session2.EXPECT().Write(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(errs[len(errs)-1])
+
+	stores[0] = local.NewStorage(session1, "metrics", resolver.NewStaticResolver(policy.NewStoragePolicy(time.Second, xtime.Second, time.Hour*48)))
+	stores[1] = local.NewStorage(session2, "metrics", resolver.NewStaticResolver(policy.NewStoragePolicy(time.Second, xtime.Second, time.Hour*48)))
+	store := NewStorage(stores, filterFunc(output))
+	return store
+}
+
+func TestFanoutReadEmpty(t *testing.T) {
+	logging.InitWithCores(nil)
```
Comment (arnikola) [inline storage/fanout/storage_test.go]: put logging init into setup methods?

File: util/execution/orderedParallel_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+func TestOrderedParallel(t *testing.T) {
+	requests := make([]Request, 3)
+	requests[0] = &request{0}
+	requests[1] = &request{1}
+	requests[2] = &request{2}
+
+	responses := ExecuteParallel(requests)
+	assert.Equal(t, responses[0].Order, 0, "ordered response")
+
```
Comment (arnikola) [inline util/execution/orderedParallel_test.go]: lots of whitespace

File: util/execution/orderedParallel_test.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type request struct {
+	order int
```
Comment (arnikola) [inline util/execution/orderedParallel_test.go]: Maybe add `value string` or something, then in `Process()`, `f.value = fmt.Sprintf("%d", f.order)`, and check that the value is set correctly?

Don't believe it's currently checking that `Process()` actually... processes :p

File: storage/fanout/storage_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+func TestFanoutWriteSuccess(t *testing.T) {
+	logging.InitWithCores(nil)
+	store := setupFanoutWrite(t, true, nil)
+	datapoints := make(ts.Datapoints, 1)
+	datapoints[0] = &ts.Datapoint{Timestamp: time.Now(), Value: 1}
+	err := store.Write(context.TODO(), &storage.WriteQuery{
+		Datapoints: datapoints,
+	})
+	assert.NoError(t, err, "write success")
```
Comment (arnikola) [inline storage/fanout/storage_test.go]: The messege at the end of the `assert` (is optional and) should explain why it failed, rather than what you're asserting

File: storage/fanout/storage_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+func setupFanoutWrite(t *testing.T, output bool, errs ...error) storage.Storage {
+	stores := make([]storage.Storage, 2)
+	ctrl := gomock.NewController(t)
+	session1 := client.NewMockSession(ctrl)
+	session2 := client.NewMockSession(ctrl)
+	session1.EXPECT().Write(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(errs[0])
+	session2.EXPECT().Write(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(errs[len(errs)-1])
+
+	stores[0] = local.NewStorage(session1, "metrics", resolver.NewStaticResolver(policy.NewStoragePolicy(time.Second, xtime.Second, time.Hour*48)))
```
Comment (arnikola) [inline storage/fanout/storage_test.go]: `stores := []storage.Storage{...}` ?

File: storage/fanout/storage_test.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	return store
+}
+
+func TestFanoutReadEmpty(t *testing.T) {
+	logging.InitWithCores(nil)
+	store := setupFanoutRead(t, false)
+	res, err := store.Fetch(context.TODO(), nil, nil)
+	assert.NoError(t, err, "No error")
+	require.NotNil(t, res, "Non empty result")
+	assert.Equal(t, len(res.SeriesList), 0, "No series")
```
Comment (arnikola) [inline storage/fanout/storage_test.go]: can use `assert.Len`

File: storage/fanout/storage.go [hunk 12]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
 
 func (s *fanoutStorage) Fetch(ctx context.Context, query *storage.FetchQuery, options *storage.FetchOptions) (*storage.FetchResult, error) {
 	stores := filterStores(s.stores, s.filter, query)
-	//TODO: Actual fanout
-	return stores[0].Fetch(ctx, query, options)
+	requests := make([]execution.Request, len(stores))
+	for idx, store := range stores {
```
[the remark this answers is not in the document]
Comment (nikunjgit) [reply storage/fanout/storage.go]: no real convention. Personally, if im' using the index inside an array then i used idx for clarity

File: storage/fanout/storage.go [hunk 13]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
...
+
+func handleFetchResponses(requests []execution.Request, responses []*execution.Response) (*storage.FetchResult, error) {
+	if len(requests) != len(responses) {
+		return nil, fmt.Errorf("invalid response from fetch")
+	}
+
+	seriesList := make([]*ts.Series, 0, len(requests))
+	result := &storage.FetchResult{SeriesList: seriesList, LocalOnly: true}
+	for idx, res := range responses {
+		if requests[idx].Order() != res.Order {
```
[the remark this answers is not in the document]
Comment (nikunjgit) [reply storage/fanout/storage.go]: execution.ExecuteParallel ensures that the results are ordered. I'll add a comment to handleFetchResponses to make it clear. First one is a function since we needed to override an interface. In the second one, we just need the data

File: storage/fanout/storage.go [hunk 14]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
...
+			return nil, fmt.Errorf("responses out of order for fetch")
+		}
+
+		// This type cast can be removed if we were to create a slice of *fetchRequest instead
+		fetchreq, ok := requests[idx].(*fetchRequest)
```
[the remark this answers is not in the document]
Comment (nikunjgit) [reply storage/fanout/storage.go]: I was trying to error out early. Any reason why we should type cast first if we can error out sooner ?

File: storage/fanout/storage.go [hunk 15]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
...
+		if !ok {
+			return nil, fmt.Errorf("invalid request type")
```
[the remark this answers is not in the document]
Comment (nikunjgit) [reply storage/fanout/storage.go]: will do

File: storage/fanout/storage.go [hunk 16]
```diff
@@ -19,14 +22,69 @@ func NewStorage(stores []storage.Storage, filter filter.Storage) storage.Storage
...
+		}
+
+		if res.Err != nil {
```
[the remark this answers is not in the document]
Comment (nikunjgit) [reply storage/fanout/storage.go]: moved it up

File: storage/fanout/storage.go (hunk 1 above)
Comment (nikunjgit) [reply storage/fanout/storage.go]: good catch! changed the initialization logic. Need append since I don't know the size beforehand

File: storage/fanout/storage.go (hunk 4 above)
Comment (nikunjgit) [reply storage/fanout/storage.go]: that's correct. I'm assuming writes are idempotent here.  Other option would be to take a parameter to tell what behavior to use. I'm just leaving that out for now until we figure out a use case for that.

File: storage/fanout/storage_test.go (hunk 10 above)
Comment (nikunjgit) [reply storage/fanout/storage_test.go]: that's gonna be one huge line

File: storage/fanout/storage_test.go (hunk 6 above)
Comment (nikunjgit) [reply storage/fanout/storage_test.go]: done

File: storage/fanout/storage_test.go (hunk 11 above)
Comment (nikunjgit) [reply storage/fanout/storage_test.go]: you know too much about asserts :)

File: util/execution/orderedParallel_test.go (hunk 7 above)
Comment (nikunjgit) [reply util/execution/orderedParallel_test.go]: fixed

File: storage/fanout/storage.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (arnikola) [reply storage/fanout/storage.go]: Just a side effect of the way the CR is laid out, when I read this thought that `.Order()` might do some processing, but makes sense since it just checks the internal value

File: storage/fanout/storage_test.go (hunk 10 above)
Comment (arnikola) [reply storage/fanout/storage_test.go]:
```golang
stores := []storage.Storage{
  local.NewStorage(session1, "metrics", resolver.NewStaticResolver(policy.NewStoragePolicy(time.Second, xtime.Second, time.Hour*48))),
 local.NewStorage(session2, "metrics", resolver.NewStaticResolver(policy.NewStoragePolicy(time.Second, xtime.Second, time.Hour*48)))
}
```

;)

File: util/execution/orderedParallel_test.go (hunk 8 above)
Comment (nikunjgit) [reply util/execution/orderedParallel_test.go]: there would be no way to get response if process doesn't run

File: util/execution/orderedParallel.go (hunk 2 above)
Comment (nikunjgit) [reply util/execution/orderedParallel.go]: fixed

File: util/execution/orderedParallel.go (hunk 3 above)
Comment (nikunjgit) [reply util/execution/orderedParallel.go]: sounds good

File: storage/fanout/storage.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (arnikola) [reply storage/fanout/storage.go]: We should probably establish an internal convention for stuff like this so we have a consistent code base before open source

File: storage/fanout/storage.go (hunk 4 above)
Comment (arnikola) [reply storage/fanout/storage.go]: I don't know a lot about go conventions, but I'd think you could  return a list of (store|writeRequest) / error pairs, but that might make the rest of the logic substantially harder 

I think even if writes are idempotent, this could lead to significant overhead  especially if it's a big series

File: storage/fanout/storage.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (nikunjgit) [reply storage/fanout/storage.go]: ok, idx for index ? :)

File: storage/fanout/storage.go [hunk 17]
```diff
@@ -3,30 +3,83 @@ package fanout
...
+	return handleFetchResponses(requestResponseChan)
+}
+
+func handleFetchResponses(requestResponseChan <-chan *execution.RequestResponse) (*storage.FetchResult, error) {
+	seriesList := make([]*ts.Series, 0)
+	result := &storage.FetchResult{SeriesList: seriesList, LocalOnly: true}
+	for reqResponse := range requestResponseChan {
+
+		if reqResponse.Response.Err != nil {
+			return nil, reqResponse.Response.Err
```
Comment (arnikola) [inline storage/fanout/storage.go]: Wouldn't this block the execution goroutine or is that fine?

Comment (arnikola) [reply storage/fanout/storage.go]: I think you could do something like
```golang
defer func() {
 go func() {
  for _ := range requestResponseChan {
  }
 }
}()
```
to drain the channel

File: storage/fanout/storage.go (hunk 1 above)
Comment (arnikola) [reply storage/fanout/storage.go]: !!

File: storage/fanout/storage.go [hunk 18]
```diff
@@ -3,30 +3,83 @@ package fanout
...
+	requests := make([]execution.Request, len(stores))
+	for idx, store := range stores {
+		requests[idx] = newWriteRequest(ctx, store, query)
+	}
+
+	requestResponseChan := execution.ExecuteParallel(requests)
+	// Fail on single error, assume writes are idempotent
+	for reqResponse := range requestResponseChan {
+		if reqResponse.Response.Err != nil {
+			return reqResponse.Response.Err
```
Comment (arnikola) [inline storage/fanout/storage.go]: same here perhaps

File: util/execution/parallel_test.go [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+	requests[2] = &request{2}
+
+	reqResponseChan := ExecuteParallel(requests)
+	responses := make([]*RequestResponse, 0)
+	for reqResponse := range reqResponseChan {
+		responses = append(responses, reqResponse)
+	}
+
+	assert.Len(t, responses, len(requests), "should have same number of responses as request")
+	assert.NotNil(t, responses[0].Response.Value, "value should not be nil")
```
Comment (arnikola) [inline util/execution/parallel_test.go]: Worth `assert.Equal(t, responses[0].Response.Value, responses[0].Request.order)` ?

File: storage/fanout/storage_test.go [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+	if len(response) == 0 {
+		response = []*fetchResponse{{err: fmt.Errorf("unable to get response")}}
+	}
+
+	ctrl := gomock.NewController(t)
+	session1 := client.NewMockSession(ctrl)
+	session2 := client.NewMockSession(ctrl)
+	session1.EXPECT().Fetch(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(response[0].result, response[0].err)
+	session2.EXPECT().Fetch(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(response[len(response)-1].result, response[len(response)-1].err)
+	stores := []storage.Storage{
```
Comment (arnikola) [inline storage/fanout/storage_test.go]: 👍

File: policy/filter/storage.go [hunk 21]
```diff
@@ -4,3 +4,8 @@ import "github.com/m3db/m3coordinator/storage"
 
 // Storage determines whether storage can fulfil the read query
 type Storage func(query storage.Query, store storage.Storage) bool
+
+// LocalOnly filters out all remote storages
+func LocalOnly(query storage.Query, store storage.Storage) bool {
+	return store.Type() == storage.TypeLocalDC
+}
```
Comment (benraskin92) [inline policy/filter/storage.go]: nit: Add newline here

File: storage/fanout/storage.go [hunk 22]
```diff
@@ -3,30 +3,83 @@ package fanout
...
 	"github.com/m3db/m3coordinator/storage"
+	"github.com/m3db/m3coordinator/ts"
+	"github.com/m3db/m3coordinator/util/execution"
 )
 
 type fanoutStorage struct {
-	stores []storage.Storage
-	filter filter.Storage
+	stores      []storage.Storage
+	readFilter  filter.Storage
```
[the remark this answers is not in the document]
Comment (benraskin92) [reply storage/fanout/storage.go]: +1

File: util/execution/parallel_test.go [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+
+func TestOrderedParallel(t *testing.T) {
+	requests := make([]Request, 3)
+	requests[0] = &request{order: 0}
+	requests[1] = &request{order: 1}
+	requests[2] = &request{order: 2}
+
+	err := ExecuteParallel(context.Background(), requests)
+	require.NoError(t, err, "no error during parallel execute")
+	assert.True(t, requests[0].(*request).processed, "request processed")
```
Comment (benraskin92) [inline util/execution/parallel_test.go]: Do you need to check that the other requests also got processed?

File: util/execution/parallel_test.go [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	if f.order == 0 {
+		time.Sleep(2 * time.Millisecond)
+	}
+
+	return nil
+}
+
+func (f *request) String() string {
```
Comment (benraskin92) [inline util/execution/parallel_test.go]: Do you need this func?

File: services/m3coordinator/main/main.go [hunk 25]
```diff
@@ -66,8 +69,12 @@ func main() {
 		os.Exit(1)
 	}
 
-	storage := local.NewStorage(session, namespace, resolver.NewStaticResolver(policy.NewStoragePolicy(time.Second, xtime.Second, time.Hour*48)))
-	handler, err := httpd.NewHandler(storage, executor.NewEngine(storage))
+	// setup storages
```
Comment (arnikola) [inline services/m3coordinator/main/main.go]: Maybe pull this out into a function? Our main is getting a bit bloated

File: storage/fanout/storage.go [hunk 26]
```diff
@@ -42,3 +83,44 @@ func filterStores(stores []storage.Storage, filterPolicy filter.Storage, query s
...
+
+func newFetchRequest(store storage.Storage, query *storage.FetchQuery, options *storage.FetchOptions) execution.Request {
+	return &fetchRequest{
+		store:   store,
+		query:   query,
+		options: options,
+	}
+}
+
+func (f *fetchRequest) Process(ctx context.Context) error {
```
Comment (benraskin92) [inline storage/fanout/storage.go]: Maybe rename to `Execute` instead of `Process`?

File: storage/fanout/storage.go [hunk 27]
```diff
@@ -25,40 +25,36 @@ func (s *fanoutStorage) Fetch(ctx context.Context, query *storage.FetchQuery, op
...
-		}
-
 		if fetchreq.store.Type() != storage.TypeLocalDC {
 			result.LocalOnly = false
 		}
 
-		result.SeriesList = append(result.SeriesList, fetchResult.SeriesList...)
+		if fetchreq.result == nil {
+			return nil, errors.ErrInvalidFetchResult
+
```
Comment (arnikola) [inline storage/fanout/storage.go]: whitespace

File: storage/fanout/storage.go [hunk 28]
```diff
@@ -25,40 +25,36 @@ func (s *fanoutStorage) Fetch(ctx context.Context, query *storage.FetchQuery, op
...
+	if err != nil {
+		return nil, err
+	}
+
+	return handleFetchResponses(requests)
 }
 
-func handleFetchResponses(requestResponseChan <-chan *execution.RequestResponse) (*storage.FetchResult, error) {
+func handleFetchResponses(requests []execution.Request) (*storage.FetchResult, error) {
 	seriesList := make([]*ts.Series, 0)
```
Comment (arnikola) [inline storage/fanout/storage.go]: Set initial size to `len(requests)` to reduce resizing in the append?

File: util/execution/parallel_test.go (hunk 23 above)
Comment (nikunjgit) [reply util/execution/parallel_test.go]: first request is the slowest one.

File: storage/fanout/storage.go (hunk 27 above)
Comment (arnikola) [reply storage/fanout/storage.go]: Also if it's erroring out, maybe do it before the previous check?

File: util/execution/parallel_test.go (hunk 24 above)
Comment (nikunjgit) [reply util/execution/parallel_test.go]: nice for debugging :)

File: util/execution/parallel.go [hunk 29]
```diff
@@ -22,22 +24,21 @@ type Response struct {
...
+func processParallel(ctx context.Context, requests []Request) error {
+	g, ctx := errgroup.WithContext(ctx)
 	for _, req := range requests {
-		go func(r Request) {
-			defer wg.Done()
-			response := r.Process()
-			requestResponse <- &RequestResponse{Request: r, Response: response}
+		// Need to use a separate func since g.Go doesn't take input
+		func (req Request) {
+			g.Go(func() error {
```
Comment (arnikola) [inline util/execution/parallel.go]: I'm guessing this thing updates the group count before it actually sets off the goroutine?

Comment (nikunjgit) [reply util/execution/parallel.go]: yea

File: storage/fanout/storage.go (hunk 4 above)
Comment (nikunjgit) [reply storage/fanout/storage.go]: changed the logic now
